pineappl/src/boc.rs

```diff
@@ -209,6 +212,460 @@ impl Scales {
+pub struct BinsWithFillLimits {
+    bins: Vec<Bin>,
+    fill_limits: Vec<f64>,
+}
+
+impl BinsWithFillLimits {
+    /// TODO
+    ///
+    /// # Errors
+    ///
+    /// TODO
+    pub fn new(bins: Vec<Bin>, fill_limits: Vec<f64>) -> Result<Self, ()> {
+        // TODO: validate the bins
+
+        // - there must be at least one bin
+        // - all fill limits must be ascending
+        // - all dimensions must be the same
+        // - limits must not overlap
+
+        if fill_limits.len() != bins.len() + 1 {
+            // TODO: do proper error handling
+            return Err(());
+        }
+
+        Ok(Self { bins, fill_limits })
+    }
+
+    /// TODO
+    ///
+    /// # Errors
+    ///
+    /// TODO
+    pub fn from_fill_limits(fill_limits: Vec<f64>) -> Result<Self, ()> {
+        let bins = fill_limits
+            .windows(2)
+            .map(|win| Bin::new(vec![(win[0], win[1])], win[1] - win[0]))
+            .collect();
+
+        Self::new(bins, fill_limits)
+    }
+
+    /// TODO
+    ///
+    /// # Errors
+    ///
+    /// TODO
+    ///
+    /// # Panics
+    ///
+    /// TODO
+    pub fn from_limits_and_normalizations(
+        limits: Vec<Vec<(f64, f64)>>,
+        normalizations: Vec<f64>,
+    ) -> Result<Self, ()> {
+        assert_eq!(limits.len(), normalizations.len());
+
+        let fill_limits = (0..=limits.len()).map(convert::f64_from_usize).collect();
+        let bins = limits
+            .into_iter()
+            .zip(normalizations)
+            .map(|(limits, normalization)| Bin::new(limits, normalization))
+            .collect();
+
+        Self::new(bins, fill_limits)
+    }
+
+    /// TODO
+    pub fn slices(&self) -> Vec<Range<usize>> {
+        if self.dimensions() == 1 {
+            // TODO: check that bins are contiguous
+            vec![0..self.len()]
+        } else {
+            self.bins()
+                .iter()
+                .flat_map(Bin::limits)
+                .enumerate()
+                .filter_map(|(index, x)| {
+                    ((index % self.dimensions()) != (self.dimensions() - 1)).then_some(x)
+                })
+                .collect::<Vec<_>>()
+                .chunks_exact(self.dimensions() - 1)
+                .enumerate()
+                .dedup_by_with_count(|(_, x), (_, y)| x == y)
+                .map(|(count, (index, _))| index..index + count)
+                .collect()
+        }
+    }
+
+    /// TODO
+    #[must_use]
+    pub fn bins(&self) -> &[Bin] {
+        &self.bins
+    }
+
+    /// TODO
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.bins.len()
+    }
+
+    /// TODO
+    #[must_use]
+    pub fn dimensions(&self) -> usize {
+        self.bins
+            .first()
+            // UNWRAP: `Bin::new` should guarantee that there's at least one bin
+            .unwrap_or_else(|| unreachable!())
+            .dimensions()
+    }
+
+    /// TODO
+    #[must_use]
+    pub fn fill_index(&self, value: f64) -> Option<usize> {
```

Comment:
this prefix is a bit confusing, i.e. `fill_index` and `fill_limits` are not similar to each other - maybe `insert_at_index`?

Reply:
or actually, just `insert`?

Reply:
They are similar to each other in that they both refer to `Grid::fill`. This method has a parameter `observable` which is an `f64`, which is being checked if it falls inside the limits returned by `BinsWithFillLimits::fill_limits`. If that's not the case, `BinsWithFillLimits::fill_index` returns `None`, else `Some(index)` where `index` is the result. Of course, this should be part of the documentation that is still missing.

Reply:
okay, I misunderstood (so yes documentation) ... mmm I still think it's not a perfect name (as usual this is a subjective thing); how about `search_fill_index`? if you disagree please just resolve this comment